docs/proposals/server-side-apply-field-managers.md

```diff
@@ -0,0 +1,283 @@
+# Server-Side Apply Dynamic Field Managers
+
+This proposal is a follow up on the [original Server-Side Apply proposal](./server-side-apply.md), and seeks to make Argo more flexible and give users the ability to apply changes to shared Kuberentes resources at a granular level. It also considers how to remove field-level changes when an Application is destroyed.
+
+To quote the [Kubernetes docs][1]:
+
+> Server-Side Apply helps users and controllers manage their resources through declarative configuration. Clients can create and modify objects declaratively by submitting their _fully specified intent_.
+
+> A fully specified intent is a partial object that only includes the fields and values for which the user has an opinion. That intent either creates a new object (using default values for unspecified fields), or is combined, by the API server, with the existing object.
+
+## Example
+
+Consider the case where you're working on a new feature for your app. You'd like to review these changes without disturbing your staging environment, so you use an `ApplicationSet` with a pull request generator to create [review apps][3] dynamically. These review apps configure a central Consul `IngressGateway`. Each review app needs to add a service to the `IngressGateway` upon creation, and remove that service upon deletion.
+
+```yaml
+apiVersion: consul.hashicorp.com/v1alpha1
+kind: IngressGateway
+metadata:
+  name: review-app-ingress-gateway
+  namespace: default
+spec:
+  listeners:
+  - port: 443
+    protocol: http
+    services:
+    - name: review-app-1
+      namespace: my-cool-app
+    - name: review-app-2
+      namespace: my-groovy-app
+    - name: review-app-3
+      namespace: my-incredible-app
+```
+
+---
+
+## Open Questions
+
+### [Q-1] What should the behavior be for a server-side applied resource upon Application deletion?
+
+The current behavior is to delete all objects "owned" by the Application. A user could choose to leave the resource behind with `Prune=false`, but that would also leave behind any config added to the shared object. Should the default delete behavior for server-side applied resources be to remove any fields that match that Application's [field manager][2]?
```

Comment:
I think this becomes very difficult to get right.

Let's imagine following existing YAML in the cluster:

```yaml
spec:
  foo: bar
  bar: baz
  baz: foo
```

Now, Argo CD comes to apply a partial patch to that resource, updating the `.spec.foo` field so that the resource now looks like the following:

```yaml
spec:
  foo: barbar
  bar: baz
  baz: foo
```

Then, another update will be made by Argo CD so that the resource will look like the following:

```yaml
spec:
  foo: barbarbar
  bar: baz
  baz: foo
```

How do we get back to the initial state, i.e. how can we properly unpatch that resource? Any last applied information will have `.spec.foo` to be `barbar`, while in reality, it should be `bar` after deleting the application.

We could just remove `.spec.foo` completely, but then we could also end up with either a resource that does not look like it should or in the worst case, would be invalid because `.spec.foo` is a required field.

Do we have any strategies for this case?

Reply:
FWIW, my team's use case for this doesn't involve multiple applications fighting for ownership of the same field. Just sibling fields on a shared resource. I imagine that'll be most common. I don't think we should support returning the resource to its original state. We're just exposing more flexibility offered by server-side apply.

If the application that originally wrote `foo: bar` wants that value back, it would only need to sync again. And it should have shown "Out of Sync" after the second application wrote `foo: barbar`.

If the system that wrote `foo: bar` existed outside of ArgoCD, I don't think that's in scope for us to support 🤷‍♂️ .

Reply:
> I don't think we should support returning the resource to its original state. We're just exposing more flexibility offered by server-side apply.

I think you're wrong in your assumption here. One of the use cases for server side apply is to be able to patch a preexisting resource on the cluster, no matter who or which system created it in the first place.

The current behavior, for a lack of a better one, is to delete the whole resource (as you wrote). And we probably should change that. Removing a field from the resource could potentially break that resource (if we patched a preexisting one). So returning a resource to the state before Argo CD came around would be a great thing to have.

That being said, I guess removing only the managed fields (unless Prune=false) is better than removing the whole resource in the first place. I was wondering what happens when a mandatory field gets deleted, though - will the delete call fail, or will the field be deleted?

Reply:
I _think_ Kubernetes would reject any operation that violated the CRD schema or failed a validating admission webhook.

> Removing a field from the resource could potentially break that resource (if we patched a preexisting one). So returning a resource to the state before Argo CD came around would be a great thing to have.

If we take ArgoCD out of the picture, and just server-side patched an existing field on a resource with `kubectl`, there would be no way to restore it without knowing the original value.

In its current state, can ArgoCD wipe out existing resources? If so, why remember a history of previous state(s) to restore upon app deletion only for server side apply?

Considering that:

> [Argo CD is largely stateless.](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/)

We'd need to store the original state in an annotation. This has size limitations, but would cover most cases.
https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#replace-resource-instead-of-applying-changes

But above all else, I selfishly want to keep this proposal simple because I'm probably going to end up implementing it 😛